Feature: Added renderTo prop instead of appendToBody

pravton opened the pull request:
<!--
Use the following format as PR title:

[docs/hooks] fix/release/feature/patch: title of pr
-->

## Description

This PR addresses issue #20 by removing the appendToBody prop and introducing the renderTo prop for the Modal component.

## Solution

- Added the renderTo prop, accepting both a string selector or an HTMLElement.
- Removed the appendToBody prop.
- If renderTo is a string, the modal will be appended to the element selected by document.querySelector. If it’s an HTMLElement, the modal will render in that element. If renderTo is not provided or fails, the modal defaults to document.body.

## Additional Notes

This is a breaking change.

## Screenshots

https://github.com/wethegit/react-modal/assets/80900245/5a692973-2e86-4c64-8ec6-346d2a537b4b

andrewrubin commented on src/lib/components/modal/modal.tsx:
```diff
@@ -6,29 +6,42 @@ import type { ModalInnerProps } from "../modal-inner"
+const getModalRoot = (renderTo?: string | HTMLElement) => {
+  if (typeof window !== "undefined") {
+    if (typeof renderTo === "string") {
+      return document.querySelector(renderTo) || document.body
```
I think we need to add some error handling here. If a user is passing an invalid selector, they won't know why their modal is being appended to the body. Maybe we just error instead of falling back to `document.body`?

pravton replied:
Oh yes, I guess an error would be more helpful than appending body, I will add some error handling here.

rvno replied:
@pravton I think there may've been a misunderstanding here based on what I'm seeing from the example.
If you tweak the `renderTo` examples where there's a string value passed in, the modal still opens when I believe what @andrewrubin meant is that if the element is not found, **no** modal should be opened.

pravton replied:
Thanks Harvey! The intension was to keep the render in-place regardless since we don't use typescript in most cases, but based on Marlon's suggestion, we should be good once we remove the string/query selector.

andrewrubin commented on src/lib/components/modal/modal.tsx:
```diff
@@ -6,29 +6,42 @@ import type { ModalInnerProps } from "../modal-inner"
+
+export function Modal({ renderTo, className, ...props }: ModalProps) {
+  const modalRoot = getModalRoot(renderTo)
+  const appendToBody = modalRoot ? modalRoot === document.body : false
```
I know modalRoot does a check, but `document.body` will always be undefined on SSR here. Could be part of the hydration issue you were describing.

I think we could also rename this `appendToBody` variable to be a little more indicative of what its purpose is (being used to determine whether the body locks scrolling)

pravton replied:
Hmm yes, but when I tried to return null based on the client side logic, based on this https://nextjs.org/docs/messages/react-hydration-error because "there was a difference between the React tree that was pre-rendered from the server"

and that make sense, I will rename the appendToBody..

andrewrubin commented on src/lib/components/modal/modal.tsx:
```diff
@@ -6,29 +6,42 @@ import type { ModalInnerProps } from "../modal-inner"
 }
 
-export function Modal({ appendToBody, className, ...props }: ModalProps) {
+const getModalRoot = (renderTo?: string | HTMLElement) => {
```
It might be worth memoizing this value as `modalRoot` within the component itself, rather than declaring a helper function outside.

pravton replied:
Fore sure, i will memorize this function within the component.

andrewrubin reviewed (changes requested):
Thanks @pravton!

Couple comments from me here, and some ideas which might help to solve the hydration issues we were discussing.

Let me know if any of my feedback isn't clear 🙏

pravton commented:
@andrewrubin I have updated based on your suggestions, let me know if this works or if anything needs to be addressed or any other suggestions.

marlonmarcello commented on src/lib/components/modal/modal.tsx:
```diff
@@ -6,29 +6,46 @@ import type { ModalInnerProps } from "../modal-inner"
 import { classnames } from "../../../utils/classnames"
```
Let's add `use client` to the top here.

pravton replied:
For sure, I will add it!

marlonmarcello commented on src/lib/components/modal/modal.tsx:
```diff
@@ -6,29 +6,46 @@ import type { ModalInnerProps } from "../modal-inner"
+  const modalRoot = useMemo(() => {
+    if (!isClient) return null
+
+    if (typeof renderTo === "string") {
```
I think we should instead remove the option to pass a `string` and only accept an `HTMLElement`.  It makes this more intention as the user will have to handle nullish values.

It also feel more reacty that way as you can use focus more on using refs.

pravton replied:
Hmm, yeah that makes sense! I will refactor this to only accept HTML element and I agree, feels more reacty this way..

marlonmarcello reviewed (changes requested):
I think we should **NOT** default to `document.body` and make that parameter mandatory instead of optional. Rendering in place by default.

Also following my other comment, we shouldn't accept a string for this I think, just an `HTMLElement`.

We should also make `positon: fixed` default regardless and let the user handle a differerent position strategy if they want.

Rendering in place with `position: fixed` covers 90% of the use cases and avoids all these SSR issues.

These changes force this prop to be more intentional overall.

pravton commented:
@andrewrubin @marlonmarcello @rvno So I made few changes based on the comments above,

- Removed string prop, and modal component now only accepts HTMLElement. I have also made this prop mandatory in ModalProps.
- By default the modal will be rendered in place with `position: fixed`
- Updated the documentation to reflect the new changes

andrewrubin commented on src/main.tsx:
I think it would be beneficial to show an example in `main.tsx` of appending a modal to the body, thoughts?

pravton replied:
For sure, yeah that would be helpful! I will add an example.

andrewrubin commented on src/main.tsx:
I know it's not part of this PR, but can we make a tweak to the last example here:

`Open modal from anchor without events` — it is unclear what this means. Can we just change the text to be more informative. i.e. "Open modal from an existing modal hash" or something. Whatever you think makes the most sense.

pravton replied:
Cool, yeah! I will update the text.

andrewrubin reviewed (changes requested):
These updates are great @pravton 👏  I think it solves all the issues nicely. Couple final ideas/questions from me.

pravton commented:
@andrewrubin So I have added an example on on the main appending the modal to body and updated the text as I see fit, but let me know if that needs to be updated or the example.

marlonmarcello commented:
Great work @pravton !

andrewrubin commented:
okay! Looking great @pravton 

Next step here is to bump the version using the `yarn changeset` command, and in the questionnaire just specify that it's a `major` version bump. I'd also like to get a `beta` release going so we can test this before releasing.

It looks like the docs for that are here: https://github.com/changesets/changesets/blob/main/docs/prereleases.md, but I'm tagging @marlonmarcello to see if he has some quick tips for us on this.

andrewrubin commented:
Thanks for the version bump @pravton 

Let's keep this open for now, we'll just need to publish a beta version before merging this, so that we can test. I think the easiest option is to do that locally. Will make a note to do this when I have a moment!

andrewrubin reviewed (approved):
@pravton , sorry mate, I didn't realize this PR was to a release candidate. We're good to merge this!

After, would you mind making a PR to main, for the release candidate? I will do the beta release from there
